Multi-team post about moving rails apps off webpacker (#19)
pulibrary/pulibrary.github.io: _posts/moving-rails-apps-off-webpacker.md

It is thankfully time to excise webpacker from all our Rails applications, and there are a lot of options for which tools to adopt.

For more background, see
* [DHH's summary of JavaScript in Rails 7](https://world.hey.com/dhh/rails-7-will-have-three-great-answers-to-javascript-in-2021-8d68191b)
* The [Rails 7 release announcement](https://rubyonrails.org/2021/12/15/Rails-7-fulfilling-a-vision)
* Rails Guides on [Working with Javascript](https://guides.rubyonrails.org/working_with_javascript_in_rails.html) and the [Asset Pipeline](https://guides.rubyonrails.org/asset_pipeline.html)

Our Rails apps at PUL are not running on Rails 7 yet but we have been moving off
of webpacker in preparation for this upgrade, and because webpacker is no longer
supported. A number of us have tried a variety of options on a variety of apps,
and we summarize our experiments and reflections here.

### Vite

On the Digital Library Services (DLS) team we've landed on Vite as a nice option
for serving both javascript and CSS. It's similar to jsbundling-rails but has
more features, and admittedly a smaller adoption footprint. We tried migrating one app to

> jsbundling-rails and then to vite. They are very similar under the hood, and the

Reviewer: ```suggestion
jsbundling-rails and then to Vite. They are very similar under the hood, and the
```

With the help of [Vite's migration guide](https://vite-ruby.netlify.app/guide/migration.html#webpacker-%F0%9F%93%A6), this was a smooth transition.

On the Research Data and Scholarship Services (RDSS) team we utilized Vite after DLS and DACS each converted one of thier repositories. [pdc_describe](https://github.com/pulibrary/pdc_describe/pull/397/files) was converted to utilize vite and the transition went smoothly.

#### Vite tips / lessons learned

- Learn to pronounce it!: https://vitejs.dev/guide/#overview
- Vite does not allow for (or need) testing assets to be served from the [development server](https://github.com/pulibrary/pdc_describe/pull/403).
- If you import any files that have a file extension other than `.js`, you will need to include
the file extension when importing it.  We had a few files with an `.es6` file extension, so we
had to change this:

      import BookCoverManager from '../orangelight/book_covers'

    to this:

      import BookCoverManager from '../orangelight/book_covers.es6'

#### But what's it like to work with?

Migration is only the first step, right? On DLS, we haven't had a work cycle yet

> with the app we moved to vite so we can't speak much to the developer

Reviewer: ```suggestion
with the app we moved to Vite so we can't speak much to the developer
```

experience. It is straightforward to run `bin/vite dev` and have autoreloading
if you don't want to refresh your browser to see changes.

> On DACS, we have made some small changes to Orangelight's javascript since moving to vite,

Reviewer: ```suggestion
On DACS, we have made some small changes to Orangelight's javascript since moving to Vite,
```

and haven't noticed much difference from our previous experience with webpacker.

### Importmaps

> Importmaps is supposed to allow you to manage your javascript dependencies

Reviewer: ```suggestion
Another option for replacing webpacker is importmaps. Importmaps is supposed to allow you to manage your javascript dependencies
```

without having node installed. It's the simplest of the new options so on DLS we
tried it with our simplest rails app. Like all our other apps, this one uses our
vue.js-based design system, [lux](https://github.com/pulibrary/lux). We got
pretty far with importmaps (and dartsass-rails), but we weren't able to get this
integration with our design system. My speculation is that lux would need to
provide a node package that bundles the full source for all its dependencies? This is partially based on a [DHH
post](https://discuss.rubyonrails.org/t/rails-es6-based-replacement-for-webpacker/78656/7)
where he says, "I’d love to see a method that downloads the pinned dependencies
at some point, but it’s not trivial for packages that have nested dependencies,
which haven’t been bundled into a single dist file."

### Jsbundling-rails with esbuild

> DLS tested this migration on Figgy, our largest app. The initial upgrades went smoothly and there was lots of primary and secondary documentation online. The configuration ended up feeling awkward and we decided to use Vite instead.

Reviewer: ```suggestion
A third option for replacing webpacker is jsbundling-rails with esbuild. DLS tested this migration on Figgy, our largest app. The initial upgrades went smoothly and there was lots of primary and secondary documentation online. The configuration ended up feeling awkward and we decided to use Vite instead.
```

### Shakapacker

> We tried shakapacker in [pdc_discovery](https://github.com/pulibrary/pdc_discovery/pull/307).  It went relatively smoothly, but we never decided to put it into production.

Reviewer: ```suggestion
The last option we explored for replacing webpacker is shakapacker. We tried shakapacker in [pdc_discovery](https://github.com/pulibrary/pdc_discovery/pull/307).  It went relatively smoothly, but we never decided to put it into production.
```

### General tips / lessons learned

> - When loading both local javascript and Vue.js components, you have to be careful
>   about how you order them. Vue.js takes over the eventing system so once it's

Reviewer: Does this concern apply to all the replacement options? If so, I'd add an intro comment like:
```suggestion
- Whichever webpacker replacement you choose, you have to be careful about how you order local javascript
and Vue.js components when loading both. Vue.js takes over the eventing system so once it's
```

had to change this:

      import BookCoverManager from '../orangelight/book_covers'

    to this:

      import BookCoverManager from '../orangelight/book_covers.es6'

#### But what's it like to work with?

Migration is only the first step, right? On DLS, we haven't had a work cycle yet
with the app we moved to vite so we can't speak much to the developer
experience. It is straightforward to run `bin/vite dev` and have autoreloading
if you don't want to refresh your browser to see changes.

On DACS, we have made some small changes to Orangelight's javascript since moving to vite,
and haven't noticed much difference from our previous experience with webpacker.

### Importmaps

Importmaps is supposed to allow you to manage your javascript dependencies
without having node installed. It's the simplest of the new options so on DLS we
tried it with our simplest rails app. Like all our other apps, this one uses our
vue.js-based design system, [lux](https://github.com/pulibrary/lux). We got
pretty far with importmaps (and dartsass-rails), but we weren't able to get this
integration with our design system. My speculation is that lux would need to

> provide a node package that bundles all its dependencies. This is partially

Reviewer 2: LUX does provide a node package that bundles all its dependencies. That's how it is distributed via npm. Does it need to provide the actual source code of all its dependencies? How would that work with dependencies that aren't pinned?

Author: I think so. I don't know exactly :(

Author: @sdellis do you think I should change this language in some way?

Author: I don't know what a dist file is, really.

Author: I updated this.
